Implement autocomplete for 'start tracking'

elbenfreund opened the pull request:
**DO NOT MERGE**

@jtojnar This is my work on autocompletion so far, please have a look if you got a minute and let me know what you think. Once we feel this is the right direction I will try to add some more tests.

This commit adds basic auto completion support to the 'raw_fact entry' of
our ``StartTrackingSreeen``.
For this we make use of ``Gtk.EntryCompletion``. An additional
difficulty is the fact that our raw string may contain 'non completion
relevant' information such as time ranges or descriptions whist the
completion should be limited to ``activity.name``/``activity.category.name``
combinations. By default ``Gtk.EntryCompletion`` substitutes the
entire entry string with the found match. This would mean we loose the
additional information.
- In order to avoid this we first create a 'temporary fact' that contains
all the parsed information present in the entry string so far.
- Then we take any activity/category related information in there to match
against our existing activities.
- Finally we use any matches to feed them back into our 'temporary
fact'.
- If a match was found, a serialized version of the 'tmp fact' gets
placed in the entry field.

Some things are worth noting:
- Due to the fact that we employ ``Fact.create_from_raw_fact`` we apply
our whole 'datetime completion' machinery to the user provided string.
Once the fact gets "written back into the entry" this means the user
provided (potentially partial) information is replaced by the fully
completed datetime information. Whist this was not our original goal it
is actually benefitial as it gives the user clear feedback about what
times will be used. it also fits with the "autocomplete theme".
- For now we provide our own 'serialize fact' helper method. This should
be obsolete once
[LIB-216](https://projecthamster.atlassian.net/browse/LIB-216) has been implemented.

jtojnar commented on hamster_gtk/helpers.py:
```diff
@@ -135,3 +138,39 @@ def get_config_instance(fallback_config_instance, app_name, file_name):
+
+
+def serialize_activity(activity):
+    """Provide a serialised activity string representation."""
```
In the comments you use ‘serialised’ while the rest of the app uses American spelling. This should probably be unified.

jtojnar commented on hamster_gtk/helpers.py:
```diff
@@ -135,3 +138,39 @@ def get_config_instance(fallback_config_instance, app_name, file_name):
+
+
+def serialize_fact(fact):
+    """Provide a serialised fact string representation."""
```
In the comments you use ‘serialised’ while the rest of the app uses American spelling. This should probably be unified.

jtojnar commented on hamster_gtk/tracking/screens.py:
```diff
@@ -166,7 +168,15 @@ def __init__(self, controller, *args, **kwargs):
         super(StartTrackingBox, self).__init__(orientation=Gtk.Orientation.VERTICAL,
                                                spacing=10, *args, **kwargs)
+        # Parsed entry string will store (potentially incomplete) ``Fact``
+        # instances in here so they can be refered to by ``ob_match_selected``.
```
Misspelling ‘refered’

jtojnar commented on hamster_gtk/tracking/screens.py:
```diff
@@ -166,7 +168,15 @@ def __init__(self, controller, *args, **kwargs):
                                                spacing=10, *args, **kwargs)
+        # Parsed entry string will store (potentially incomplete) ``Fact``
+        # instances in here so they can be refered to by ``ob_match_selected``.
+        # It's main use is to store information that is part of the user
```
Misspelling ‘It’s’

jtojnar commented on hamster_gtk/tracking/screens.py:
```diff
@@ -178,13 +188,105 @@ def __init__(self, controller, *args, **kwargs):
+            """
+            Quick and dirty function to remove duplicate list items.
+
+            The first item encounterd is keept, the following thrown away.
```
Misspelling ‘encounterd’ and ‘keept’

jtojnar commented on hamster_gtk/tracking/screens.py:
```diff
@@ -178,13 +188,105 @@ def __init__(self, controller, *args, **kwargs):
+
+            The first item encounterd is keept, the following thrown away.
+
+            We can not simply turn this into a set as we would loose the
```
Misspelling ‘loose’

jtojnar commented on hamster_gtk/tracking/screens.py:
```diff
@@ -178,13 +188,105 @@ def __init__(self, controller, *args, **kwargs):
+            The first item encounterd is keept, the following thrown away.
+
+            We can not simply turn this into a set as we would loose the
+            ordering.  This clearly is highly inefficiennt and the only reason
```
Two spaces between ‘ordering.’ and ‘This’; misspelling ‘inefficiennt’

jtojnar commented on hamster_gtk/tracking/screens.py:
```diff
@@ -178,13 +188,105 @@ def __init__(self, controller, *args, **kwargs):
+            ordering.  This clearly is highly inefficiennt and the only reason
+            we do this for now is the very limited size of the input list.
+
+            For one possible way to do this see: ``https://code.activestate.com/recipes/576694/``.
```
There is an efficient OrderedSet implementation on PyPi: https://pypi.python.org/pypi/orderedset

jtojnar commented on hamster_gtk/tracking/screens.py:
```diff
@@ -178,13 +188,105 @@ def __init__(self, controller, *args, **kwargs):
+
+        def match_anywhere(completion, entrystr, iter, data):
+            """
+            Check if the entrysring is a substring of our text_column.
```
Misspelling ‘entrysring’

jtojnar commented on hamster_gtk/tracking/screens.py:
```diff
@@ -178,13 +188,105 @@ def __init__(self, controller, *args, **kwargs):
+                completion (Gtk.EntryCompletion): The completion instance.
+                entrystr (str): The text extracted from the entry so far.
+                iter: Iterator position indicating the model row we check against.
+                data: ???.
```
This is the `None` passed to the `set_match_func` method: https://lazka.github.io/pgi-docs/#Gtk-3.0/callbacks.html#Gtk.EntryCompletionMatchFunc

jtojnar commented:
1. The completion does not work for category-less activities:
   
   ```
   Traceback (most recent call last):
     File "~/projects/hamster-gtk/hamster_gtk/tracking/screens.py", line 260, in on_match_selected
       fact.activity.category = Category(category_name)
     File "~/.virtualenvs/hamster-gtk/lib/python3.5/site-packages/hamster_lib/objects.py", line 51, in __init__
       self.name = name
     File "~/.virtualenvs/hamster-gtk/lib/python3.5/site-packages/hamster_lib/objects.py", line 61, in name
       raise ValueError(_("You need to specify a name."))
   ValueError: You need to specify a name.
   ```
2. The RawFactWidget should be factored out to a separate class which will encapsulate the completion. This will also allow to use it for both creating and editing the facts.
3. What is the `tracking-canceled` event for? Why is it part of this PR?
4. Should this only suggest the activities from today? I do some activities only once a week and I would still like them to be suggested. Not sure where to draw the line though – one can as well argue for once a year and the suggestion would be even more useful there as the user won’t probably remember the exact name.

jtojnar commented:
Entering space also produces an error.

elbenfreund commented:
what do you mean with "entering space"?

jtojnar commented:
Entering space characters using space bar.

elbenfreund commented:
Do you mean the error dialog that appears or do you get an unhandled exception? What is your expected behaviour?

jtojnar commented:
Unhandled error:

```
ValueError: You need to specify a name.
Traceback (most recent call last):
  File "~/hamster-gtk/hamster_gtk/tracking/screens.py", line 243, in match_anywhere
    fact = Fact.create_from_raw_fact(entrystr)
  File "~/hamster-gtk/lib/python3.5/site-packages/hamster_lib/objects.py", line 478, in create_from_raw_fact
    activity = Activity(activity_name, category=category)
  File "~/hamster-gtk/lib/python3.5/site-packages/hamster_lib/objects.py", line 147, in __init__
    self.name = name
  File "~/.virtualenvs/hamster-gtk/lib/python3.5/site-packages/hamster_lib/objects.py", line 159, in name
    raise ValueError(_("You need to specify a name."))
```

Expected behaviour would be something like this:

``` python
if entrystr.strip() == '':
    return True
```

elbenfreund committed:
Add custom 'RawFactEntry'

This commit provides a new custom entry class ``RawFactEntry`` suitable
for add/edit fact dialogs/boxes. One main feature of this new class is
that it provides basic autocomplete functionality via a custom
``Gtk.EntryCompletion`` subclass.

A major issue we had to solve was the fact that we can not simply look
up viable matches and replace the 'raw_fact' string with an activity
string because this would remove any existing time or description
informmation contained in the entry string.

Another fact worth noting is that this commit introdices serialisation
hellper methods for ``Activity`` and ``Fact`` in order to avoid code
duplication. Once ``LIB-216`` has been implemented these may no longer
be needed.

Closes: #136

jtojnar commented on hamster_gtk/misc/widgets/raw_fact_entry.py:
```diff
@@ -0,0 +1,127 @@
+    """
+
+    def __init__(self, controller, *args, **kwargs):
+        """instantiate class."""
```
In [`RawFactEntry.__init__`](https://github.com/projecthamster/hamster-gtk/pull/138/commits/b7e7d875f60314be495f0bb9d2b1b1da192b5c6d#diff-f5e2db386bd89261e3d69c4dfeda3181R35) you use ‘Instantiate the class.’ while here you just use ‘instantiate class.’ It should probably be unified, at least in the same file.

jtojnar commented on hamster_gtk/misc/widgets/raw_fact_entry.py:
```diff
@@ -0,0 +1,127 @@
+    # Callbacks
+    def _on_facts_changed(self, evtl):
+        """Callback triggered when facts have changed."""
+        self.set_completion(RawFactCompletion(self._controller))
```
Rather than creating a new completion each time, we should probably just propagate the event to the completion and let it update its own model. Or since we are already passing the controller to the completion, the completion could subscribe the `facts-changed` signal on its own.

jtojnar commented on hamster_gtk/misc/widgets/raw_fact_entry.py:
```diff
@@ -0,0 +1,127 @@
+
+    def _match_anywhere(self, entrystr, iter, data, *args):
+        """
+        Check if the entrysring is a substring of our text_column.
```
Misspelling entrysring

jtojnar commented on hamster_gtk/misc/widgets/raw_fact_entry.py:
```diff
@@ -0,0 +1,127 @@
+            data: Arbitrary user data.
+
+        Returns:
+            boot: ``True`` if ``entrystring`` is a substring, ``False`` if not.
```
Misspelling boot

jtojnar commented on hamster_gtk/misc/widgets/raw_fact_entry.py:
```diff
@@ -0,0 +1,127 @@
+            start=today, end=today)]
+        return OrderedSet(recent_activities)
+
+    def _match_anywhere(self, entrystr, iter, data, *args):
```
Where does the `args` argument come from?

https://lazka.github.io/pgi-docs/#Gtk-3.0/callbacks.html#Gtk.EntryCompletionMatchFunc has no mention of it.

jtojnar replied:
Actually you are missing the `completion` parameter (see the reference). It should be like this:

``` python
def _match_anywhere(self, completion, entrystr, iter, data):
```

It is the cause of the exception I mention in the comment bellow.

jtojnar commented on hamster_gtk/misc/widgets/raw_fact_entry.py:
```diff
@@ -0,0 +1,127 @@
+            boot: ``True`` if ``entrystring`` is a substring, ``False`` if not.
+
+        Note this is a custom match function, for details on the general generic
+        solution [please see|https://lazka.github.io/pgi-docs/#Gtk-3.0/
```
Links should not really be broken like this. Also what is the `[link|url]` syntax?

jtojnar commented on hamster_gtk/misc/widgets/raw_fact_entry.py:
```diff
@@ -0,0 +1,127 @@
+        # 'activity@category' string that we try to match against.
+        modelstr = _u(self.get_model()[iter][0])
+        # Parse the entire entry string to extract all semantic information.
+        fact = Fact.create_from_raw_fact(_u(entrystr))
```
Creating the fact for each activity in the model is not really efficient. Maybe we could subscribe for entry changes and update the fact then (and only accessing `self._fact` here. Or even `self._activity_string`.

elbenfreund replied:
I agree. However, I would have to look into the exact way signals are triggered/handled here in order to prevent race conditions between processing `text-changed` and `match_anywhere`.
In case this turns out to be non trivial it may be possible to take it as it is and open an enhancement issue instead...

jtojnar replied:
Other way is caching the fact string in a property – something like this:

``` python
cache_key = self._activity_string_cache[0]
if cache_key != entrystr:
        fact = Fact.create_from_raw_fact(_u(entrystr))
        if fact.activity:
            activity_string = helpers.serialise_activity(fact.activity)
        self._activity_string_cache = (entrystr, activity_string)
activity_string = self._activity_string_cache[1]
```

jtojnar commented on hamster_gtk/tracking/screens.py:
```diff
@@ -59,7 +60,7 @@ def update(self, evt=None):
         This depends on wether there exists an *ongoing fact* or not.
```
Misspelling wether

jtojnar commented on tests/misc/conftest.py:
```diff
@@ -62,3 +66,41 @@ def monthrange_parametrized(request):
+
+@pytest.fixture
+def activity_model(request, activity):
+    """Returnd a ``ListStore`` instance with a generic ``Activity`` as it's only row."""
```
Misspelling Returnd and it's

jtojnar commented:
The current code does not work for me. I had to create `hamster_gtk/misc/widgets/__init__.py` re-exporting `RawFactEntry` and add re-export `RawFactEntry` from `hamster_gtk/misc/__init__.py` before I managed to start the application.

I also had to install `OrderedSet` package, it should probably be added to `setup.py`.

Even then, entering any text into the field causes an error

```
(hamster-gtk:30195): Gtk-WARNING **: gtktreemodel.c:645: Invalid path test passed to gtk_tree_path_new_from_string
Traceback (most recent call last):
  File "/usr/lib/python3.5/site-packages/gi/overrides/Gtk.py", line 1162, in __new__
    return TreePath.new_from_string(path)
TypeError: constructor returned NULL

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "~/projects/hamster-gtk/hamster_gtk/misc/widgets/raw_fact_entry.py", line 100, in _match_anywhere
    modelstr = _u(self.get_model()[iter][0])
  File "/usr/lib/python3.5/site-packages/gi/overrides/Gtk.py", line 801, in __getitem__
    aiter = self._getiter(key)
  File "/usr/lib/python3.5/site-packages/gi/overrides/Gtk.py", line 789, in _getiter
    aiter = self.get_iter(key)
  File "/usr/lib/python3.5/site-packages/gi/overrides/Gtk.py", line 823, in get_iter
    path = self._coerce_path(path)
  File "/usr/lib/python3.5/site-packages/gi/overrides/Gtk.py", line 798, in _coerce_path
    return TreePath(path)
  File "/usr/lib/python3.5/site-packages/gi/overrides/Gtk.py", line 1164, in __new__
    raise TypeError("could not parse subscript '%s' as a tree path" % path)
TypeError: could not parse subscript 'test' as a tree path
```

**EDIT:** Actually, it is caused by a missing parameter in a `_match_anywhere` method. See https://github.com/projecthamster/hamster-gtk/pull/138#discussion_r87098991

jtojnar commented on hamster_gtk/misc/widgets/raw_fact_entry.py:
```diff
@@ -0,0 +1,127 @@
+            activity_string = helpers.serialise_activity(fact.activity)
+        return activity_string in modelstr
+
+    def _on_match_selected(self, model, iter):
```
Similarly to https://github.com/projecthamster/hamster-gtk/pull/138#discussion_r87098991 there should be `completion` parameter (see https://lazka.github.io/pgi-docs/#Gtk-3.0/classes/EntryCompletion.html#Gtk.EntryCompletion.signals.match_selected). Otherwise this error is raised:

``` python
TypeError: _on_match_selected() takes 3 positional arguments but 4 were given
```

elbenfreund replied:
strange. our signature provides 3 parameter spots just as in the docs you referenced. The only difference is the naming of the first one (`self` vs `entry_completion`) which the code should not care about.

jtojnar replied:
I guess, when passing a method as an argument, the object is bound to `self`, i.e. `obj.method(*args)` is actually functionally equivalent to `lambda *args: obj.method(*args)` – `obj` is captured in a closure, similarly to writing `lambda *args: getattr(obj, 'method')(*args)`

elbenfreund commented:
After some deliberation I think this approach is not all that viable. Whilst it can be made work 'sort of' it its gets hackier by the minute if one tries to have seperate activity.name and activity.category.name autocompletion let alone tags. I will cose this for now and try something else instead.

For reference: Please note that the branch name actually refers to the wrong issue (#136) which itelf is actually not the best fit to begin with.
